dashboard/design/survey_summary_design.md

```diff
@@ -0,0 +1,161 @@
+
+### Detail design
+The pipeline has 4 main actions
+1. Ingest: Download data from JotForm server to CDO database.
+2. Query: Find survey data that matches lookup conditions.
+3. Summarize: Aggregate survey data.
+4. Present: Display survey summaries in dashboard.
+
+![Diagram](survey_summary_diagram.png)
+
+Some patterns this design follows
+- _Open-Closed Principle_\
+  The pipeline is open for extension such as adding new components or adding different implementation of a component. However, it avoids, as best as it could, modifying existing code which could introduce regression. Following Single Responsibility principle will help with this.
+- _Single Responsibility Principle_\
+  Each method, class, module should have only 1 clear purpose. In a stronger statement, each should have only one reason to change.
+- _Dependency Inversion Principle_\
+  High-level modules (complex logic) should not depend on low-level modules (primary operations). Both should depend on abstractions. This allows high-level modules to easily switch to different low-level modules implementation.
+- _Template Method Pattern_\
+  Define the skeleton of an algorithm, deferring some steps to other components. E.g. Query module (step 2) could use different implementations of Parser, Indexer and Filter depends on how data is stored in database. The specific implementations of those child components will be send to Query component as input parameters (dependency injection). Strategy pattern enables switching among implementations of the same component.
+- _Strategy Pattern_\
+  Define an interface common to a family of algorithms, e.g. survey result reducers, and make them interchangeable. This lets the algorithm vary independently from clients that use it.
+
+### Build from the end
+We don't have to start from step 1 (Ingest), we can start from step 3 (Summarize) and create adapters to connect it to the current pipeline.
+
+Specifically for CSF survey work, we will create an input adapter to retrieve survey data from the current data model, and an output adapter to convert summary result to format that the current UI view understands. Looking at it from a different angle, these adapters are just concrete implementations of Retriever and Decorator.
+
+
+# Recommendation
+Option 3. Build a new generic survey pipeline, starting from step 3 (Summarize module) first and create adapters to plug it in the current pipeline. This approach allows us to support 2019 CSF survey sooner by taking advantage of the existing work. Then, gradually build out the rest of the pipeline later.
+
+## Rollout plan
+1. Build 5 components just enough to support 2019 surveys: Retriever (for current db), Transformer, Mapper, Reducer and Decorator (for current UI view). They are components in step 3 and 4 of the above diagram.
+2. Build Fetcher (step 1) to download survey submissions from JotForm to our database.
```

Comment:
I think our existing system more-or-less does exactly this - all JotForm submissions get dropped into one table.  Can we just reuse what we've got, for the Fetcher, or are there specific changes you expect we'll need?

Reply:
Yes, we can potentially reuse most of the code for this part.

One thing that I know we will have to do differently is that the current code purposefully strips away question info from submissions downloaded by `GET /form/{id}/submissions`. It later links answers back to question info downloaded from another request `GET /form/{id}/questions`. 
The problem with this is it breaks question-answer relationship, making the submission content not self-sufficient. An answer for older-version of a question could be linked to newer-version of the same question.